Fixing comment plugin not using user settings when overriding default setting

Neko-Box-Coder opened the pull request:
If you are overriding the default comment format say
(`"//%s"` instead of `"// %s"`)
```json
"*.cpp": {
    "commenttype": "//%s"
}
```

This will get ignored when switching between different file types and the default will be used again.

And also cleaned up unnecessary (and flawed) logic since we should always honor the buffer settings anyway.

dmaluka commented:
> ```json
> "*.cpp": {
>     "commenttype": "//%s"
> }
> ```

You probably meant the following?

```json
"ft:c++": {
    "commenttype": "//%s"
}
```

since with `"*.cpp"` the setting is associated with the file name, not with the file type, so it should not change after `set filetype`.

...So, with this PR and the above `"ft:c++"` setting, when I open `foo.go` and then run `set filetype c++`, the setting indeed takes effect, however, after that, when I run `set filetype go`, it doesn't restore the original setting.

Neko-Box-Coder commented:
> since with `"*.cpp"` the setting is associated with the file name, not with the file type, so it should not change after `set filetype`.

It does change when currently without this PR. It overwrites the setting from the user which is set in the beginning when opening the buffer.

> ...So, with this PR and the above `"ft:c++"` setting, when I open `foo.go` and then run `set filetype c++`, the setting indeed takes effect, however, after that, when I run `set filetype go`, it doesn't restore the original setting.

The problem here is it registers the comment type when you do a comment.

So yes... although your example is a bit unlikely, I get what you are saying. 
A more likely scenario but still introduce the problem in the same way would be:
1. the user opens a file that micro doesn't recognize, creates a new buffer or micro recognize the file type wrongly
2. then the user invoke the comment action, which registers the "wrong" comment type
3. then the user sets the correct file type and invokes the comment action, which will still continue to be wrong.

Neko-Box-Coder commented:
Actually, even for filetypes that are not registered, this bug would still manifest because of the if statement
https://github.com/zyedidia/micro/blob/21bb61c5ff3cd4cd11adbd521657df90d50f54c7/runtime/plugins/comment/comment.lua#L67-L75

JoeKar commented:
Just a thought:
Maybe `RegisterCommonOptionPlug` resp. within the plugin `RegisterCommonOption` is an option?
But again, the compatibility to `commenttype` would brake, since it's then called `comment.commenttype` or `comment.type`.

Upon `filetype` changes this option can then be cleared and the proper reaction can take place.

Neko-Box-Coder commented:
Okay, how about now @dmaluka @JoeKar 

I have added a setting called `commentfiletype` (although I am leaning towards calling it `internal_commentfiletype` to make it clear it is for internal use) which just records what file type `commenttype` is for.

It will detect file type change and apply the corresponding comment format correctly.

The new change will also update the comment format table according to the user settings for any new buffers.
So for example if you override the go comment to be 
```json
"ft:go": {
    "commenttype": "/* %s */"
}
```
for whatever reason, it will update the table so that this will be used next time.

So if you have the above json and 
1. open a .go file 
2. invoke comment (so that it registers it)
3. change the file type to something else like `setlocal filetype json`
4. invoke comment again
5. then set the filetype back to go and invoke the comment

The user settings will be applied instead of the default one.

The only niche use case this will break is if you want a specific `commenttype` for only 1 buffer and not apply to other buffers with the same file type, this will not work.

However, the original implementation doesn't work for this niche use case anyway since the moment you switch to a different buffer, your `commenttype` for that specific buffer will be gone. So it's not like this is breaking backward compatibility if it was never working :joy:

JoeKar commented:
Due to the merge of #3343 a rebase is necessary.

> I have added a setting called `commentfiletype` (although I am leaning towards calling it `internal_commentfiletype` to make it clear it is for internal use) which just records what file type `commenttype` is for.

See my https://github.com/zyedidia/micro/pull/3424#issuecomment-2287107171 above. `RegisterCommonOption()` will create/register a plugin specific option including default value, which can be used for further tracking and should be touched now (at least in theory) in case of file type changes by `micro` itself.

JoeKar commented:
> The only niche use case this will break is if you want a specific commenttype for only 1 buffer and not apply to other buffers with the same file type, this will not work.

This should do the trick for all the use cases (base is current `master`):
```
diff --git a/runtime/plugins/comment/comment.lua b/runtime/plugins/comment/comment.lua
index f86da945..50d63ee5 100644
--- a/runtime/plugins/comment/comment.lua
+++ b/runtime/plugins/comment/comment.lua
@@ -61,17 +61,15 @@ ft["zig"] = "// %s"
 ft["zscript"] = "// %s"
 ft["zsh"] = "# %s"
 
-local last_ft
-
 function updateCommentType(buf)
-    if buf.Settings["commenttype"] == nil or (last_ft ~= buf.Settings["filetype"] and last_ft ~= nil) then
+    -- NOTE: Don't use SetOptionNative() to set "comment.type",
+    -- otherwise "comment.type" can't be reset by a "filetype" change.
+    if buf.Settings["comment.type"] == "" then
         if ft[buf.Settings["filetype"]] ~= nil then
-            buf:SetOptionNative("commenttype", ft[buf.Settings["filetype"]])
+            buf.Settings["comment.type"] = ft[buf.Settings["filetype"]]
         else
-            buf:SetOptionNative("commenttype", "# %s")
+            buf.Settings["comment.type"] = "# %s"
         end
-
-        last_ft = buf.Settings["filetype"]
     end
 end
 
@@ -88,7 +86,7 @@ function commentLine(bp, lineN, indentLen)
     updateCommentType(bp.Buf)
 
     local line = bp.Buf:Line(lineN)
-    local commentType = bp.Buf.Settings["commenttype"]
+    local commentType = bp.Buf.Settings["comment.type"]
     local sel = -bp.Cursor.CurSelection
     local curpos = -bp.Cursor.Loc
     local index = string.find(commentType, "%%s") - 1
@@ -114,7 +112,7 @@ function uncommentLine(bp, lineN, commentRegex)
     updateCommentType(bp.Buf)
 
     local line = bp.Buf:Line(lineN)
-    local commentType = bp.Buf.Settings["commenttype"]
+    local commentType = bp.Buf.Settings["comment.type"]
     local sel = -bp.Cursor.CurSelection
     local curpos = -bp.Cursor.Loc
     local index = string.find(commentType, "%%s") - 1
@@ -178,7 +176,7 @@ end
 function comment(bp, args)
     updateCommentType(bp.Buf)
 
-    local commentType = bp.Buf.Settings["commenttype"]
+    local commentType = bp.Buf.Settings["comment.type"]
     local commentRegex = "^%s*" .. commentType:gsub("%%","%%%%"):gsub("%$","%$"):gsub("%)","%)"):gsub("%(","%("):gsub("%?","%?"):gsub("%*", "%*"):gsub("%-", "%-"):gsub("%.", "%."):gsub("%+", "%+"):gsub("%]", "%]"):gsub("%[", "%["):gsub("%%%%s", "(.*)")
 
     if bp.Cursor:HasSelection() then
@@ -204,6 +202,10 @@ function string.starts(String,Start)
     return string.sub(String,1,string.len(Start))==Start
 end
 
+function preinit()
+    config.RegisterCommonOption("comment", "type", "")
+end
+
 function init()
     config.MakeCommand("comment", comment, config.NoComplete)
     config.TryBindKey("Alt-/", "lua:comment.comment", false)

```

Neko-Box-Coder commented:
@JoeKar 
Thanks a lot. Yeah your approach is much better. It is working as expected now.
I also added an error message when the old option is being used to remind the user to update their settings.

Andriamanitra commented on runtime/plugins/comment/comment.lua:
```diff
@@ -61,17 +62,20 @@ ft["zig"] = "// %s"
 function updateCommentType(buf)
-    if buf.Settings["commenttype"] == nil or (last_ft ~= buf.Settings["filetype"] and last_ft ~= nil) then
+    if buf.Settings["commenttype"] ~= nil then
+        micro.InfoBar():Error("\"commenttype\" option has been updated to \"comment.type\"",
+                              "instead, please update accordingly")
```
I think this wording is harder to misinterpret:

```suggestion
        micro.InfoBar():Error("\"commenttype\" option has been renamed to \"comment.type\"",
                              ", please update your configuration")
```

dmaluka commented on runtime/plugins/comment/comment.lua:
```diff
@@ -61,17 +61,15 @@ ft["zig"] = "// %s"
 function updateCommentType(buf)
-    if buf.Settings["commenttype"] == nil or (last_ft ~= buf.Settings["filetype"] and last_ft ~= nil) then
+    -- NOTE: Don't use SetOptionNative() to set "comment.type",
+    -- otherwise "comment.type" can't be reset by a "filetype" change.
```
It is not nearly obvious *why* can't it be reset by a "filetype" change if we use `SetOptionNative()` (it took me some time to figure it out, for instance). We should explain it better, e.g. "Don't use SetOptionNative() to set "comment.type", otherwise "comment.type" will be marked as locally overridden and will not be updated when "filetype" changes."

Or actually it seems better to:
1. use `DoSetOptionNative()` instead of setting the option directly
2. add documentation to both `SetOptionNative()` and `DoSetOptionNative()`, so that everyone knows what is the difference between them and when to use each of them, and then we don't need this comment here.

...Another option would be to somehow extend `RegisterCommonOption()` to allow registering default per-filetype values of an option, not just its default global value, at init time, - so that `updateCommentType()` would not be needed at all.

Neko-Box-Coder replied:
@dmaluka 

`DoSetOptionNative()` seems like an internal function to me, the name of it is also quite confusing against `SetOptionNative()` as well.

Would adding a function like `SetOptionPersistent(bool persistent)` which sets `b.LocalSettings[option]` be enough for now?

dmaluka replied:
As matter of fact, the line between internal and external functions here is pretty blurred. (Technically `DoSetOptionNative()` is external, it is already exported to plugins, although not on purpose but as a side effect of exporting it to other modules *inside* micro, which is a consequence of the bizarre design of micro's plugin system.)

> Would adding a function like `SetOptionPersistence(bool persistent)` which sets `b.LocalSettings[option]`

Maybe... @JoeKar what do you think?

JoeKar replied:
> It is not nearly obvious _why_ can't it be reset by a "filetype" change if we use `SetOptionNative()` (it took me some time to figure it out, for instance).

I thought you can remember, after the long review road of #3343. :wink: 

> > Would adding a function like `SetOptionPersistence(bool persistent)` which sets `b.LocalSettings[option]`
>
> Maybe... @JoeKar what do you think?

So `SetOptionNative()` calls then `SetOptionNativeMark(option string, nativeValue interface{}, mark(Local) bool)`.
The same then for consistent reason for `SetGlobalOptionNative()` -> `SetGlobalOptionNativeMark(option string, nativeValue interface{}, mark(Modified) bool)`

The word `persistent` resp. `persistence` smells a bit of writing it persistent into the users configuration.

BTW: Wasn't the extension with a further parameter of these functions temporary part of #3343, but rejected due to the fact a of the introduction of a further parameter?

Anyway, if fine with that adjustment, in case it helps to improve the code/interfaces. Indeed we can then drop the introduced comment in the `comment` plugin.

dmaluka replied:
I think what @Neko-Box-Coder meant is a separate function just for toggling the option's persistence state, independently of setting the option value.

JoeKar replied:
Hm, wouldn't it over-complicate the situation (complex solution for simple problem)?
Where does `SetOptionNative()` and `DoSetOptionNative()` (current naming) not fulfill the needs to add this additional function to control the marking alone?

dmaluka replied:
I have the same thoughts.

Neko-Box-Coder replied:
Okay, that's fine. 

It's just I am not sure how to rename `DoSetOptionNative()` such that it is not confusing and doesn't require a comment explaining what the difference is.

Would it work if I add a proxy function called something like `SetNonReloadableOptionNative()` (or some other names) which does the same as `SetOptionNative()` but without setting the `LocalSettings` field.

Neko-Box-Coder replied:
It's fine to add documentation I think but it would be **difficult to rename or change** the function signature if we decide to use it in the plugin (which other people will do the same as well if needed)

I still stand by what I said before:

> It's just I am not sure how to rename `DoSetOptionNative()` such that it is not confusing and doesn't require a comment explaining what the difference is.
> 
> Would it work if I add a proxy function called something like `SetNonReloadableOptionNative()` (or some other names) which does the same as `SetOptionNative()` but without setting the `LocalSettings` field.

Or even a step further where we make `DoSetOptionNative()` as private just like `doSetGlobalOptionNative()`. The only place it is using `DoSetOptionNative()` is `internal/action/command.go:608` after all.

dmaluka replied:
I don't think `DoSetOptionNative` is a particularly bad name (but you are welcome to suggest a better one). It aligns with what it does (and what is described in the documentation I've suggested above): just sets the option for this buffer and does nothing more.

Also I hope it is unlikely that any other plugin will actually want to use it. The `comment` plugin's use case is unusual in this regard.

> Or even a step further where we make `DoSetOptionNative()` as private just like `doSetGlobalOptionNative()`. The only place it is using `DoSetOptionNative()` is `internal/action/command.go:608` after all.

`internal/action/command.go` is in a different package. Which is why we made `DoSetOptionNative()` public in the first place. Try making it private and compiling micro.

dmaluka commented on runtime/plugins/comment/comment.lua:
```diff
@@ -64,8 +64,8 @@ ft["zsh"] = "# %s"
-        micro.InfoBar():Error("\"commenttype\" option has been updated to \"comment.type\"",
-                              "instead, please update accordingly")
+        micro.InfoBar():Error("\"commenttype\" option has been renamed to \"comment.type\"",
+                              ", please update your configuration")
```
After changing `commenttype` to `comment.type` in `settings.json` and running the `reload` command, micro still stubbornly shows this error.

JoeKar commented:
@dmaluka:
Any complains?

dmaluka commented:
1. I'm not sure we should show this noisy warning about `commenttype` vs `comment.type`. Wouldn't it be friendlier to silently support `commenttype` (especially since in https://github.com/zyedidia/micro/pull/3424/commits/f12b9410a2b8fc2906efb15c50479aa781ca9134 we already support it, and BTW the warning doesn't tell the user about that), and just add a note to `comment.md` that `commenttype` is supported too but just for backward compatibility and should not be used?
2. Squash and/or split commits appropriately? (as usual, to have "one functional change per commit" rather than "introduce a problem in one commit and then fix it in another commit" or "implement a change in one commit and then reimplement it in a different way in another commit" , to avoid making review more difficult than it needs to be, to avoid  making bisection potentially more difficult than it needs to be, etc)

Neko-Box-Coder commented:
> 1. I'm not sure we should show this noisy warning about `commenttype` vs `comment.type`. Wouldn't it be friendlier to silently support `commenttype` (especially since in [f12b941](https://github.com/zyedidia/micro/commit/f12b9410a2b8fc2906efb15c50479aa781ca9134) we already support it, and BTW the warning doesn't tell the user about that), and just add a note to `comment.md` that `commenttype` is supported too but just for backward compatibility and should not be used?
> 
>     2. Squash and/or split commits appropriately? (as usual, to have "one functional change per commit" rather than "introduce a problem in one commit and then fix it in another commit" or "implement a change in one commit and then reimplement it in a different way in another commit" , to avoid making review more difficult than it needs to be, to avoid  making bisection potentially more difficult than it needs to be, etc)

Removed the warning and put it in `comment.md` instead.

There aren't that many changes anyway so I squashed them into a single commit instead.

dmaluka commented on runtime/plugins/comment/help/comment.md:
````diff
@@ -80,18 +80,23 @@ but it is only available for certain filetypes:
 }
 ```
+
+`commenttype` was the previous option name that was replaced by `comment.type`.
````
Sounds a bit unclear and torn out of some context.

How about something like: "For backward compatibility, use can also use the option name `commenttype` (without the dot) instead of `comment.type`. It is recommended to use `comment.type` instead, as `commenttype` can get deprecated in the future."

Neko-Box-Coder replied:
This is pretty minor but how about 

``` text
`commenttype` (without the dot) is the legacy option that is superseded by `comment.type`.

`commenttype` is still supported but will get deprecated in the future. 

**Use `comment.type` instead.**
```

Just firmer wording so that we can discourage people from using `commenttype`.

Andriamanitra replied:
Shouldn't it just say `commenttype` is deprecated? "Will get deprecated in the future" doesn't make much sense to me.

> Features are deprecated, rather than immediately removed, to provide backward compatibility and to give programmers time to bring affected code into compliance with the new standard.

JoeKar replied:
> Shouldn't it just say `commenttype` is deprecated?

I agree.
It is deprecated in the moment of merge.

dmaluka commented on runtime/plugins/comment/help/comment.md:
```diff
@@ -80,18 +80,22 @@ but it is only available for certain filetypes:
+
+`commenttype` (without the dot) is the legacy option that is superseded by `comment.type`.
+`commenttype` is still supported but will get deprecated in the future.
+**Use `comment.type` instead.**
```
It's better to stick to the rule of wrapping lines to not exceed a certain number of characters per line. Generally in micro's help files this limit is around 80 characters. In this particular file, as I see, it is even less (more like 65 characters), so it's better to stay consistent with that.

JoeKar commented on runtime/plugins/comment/help/comment.md:
````diff
@@ -80,18 +80,23 @@ but it is only available for certain filetypes:
 ```
+
+`commenttype` (without the dot) is the legacy option that is
+superseded by `comment.type`. `commenttype` is still supported
+but will get deprecated in the future.
````
@Neko-Box-Coder:
As mentioned [here](https://github.com/zyedidia/micro/pull/3424#discussion_r2106326952), can you adapt it to something like:
```suggestion
superseded by `comment.type`. `commenttype` is still supported
but deprecated from now on.
```

dmaluka commented on runtime/plugins/comment/help/comment.md:
```diff
@@ -80,18 +80,23 @@ but it is only available for certain filetypes:
+
+`commenttype` (without the dot) is the legacy option that is
+superseded by `comment.type`. `commenttype` is still supported
+but deprecated from now on.
```
What does "now" mean to someone who is ready this, say, 3 years from now?

Neko-Box-Coder replied:
Updated.
